### ISSUE TITLE
Support an operation to have multiple specifications per path (e.g. multiple POST operation per path)

### ISSUE THREAD

**Issue by DavidBiesack** [OAI/OpenAPI-Specification#182]
with POST actions especially, there can be several unrelated operations that would be better presented to the user as separate POST actions in swagger-ui because they do very different things to the same resource. Each of these different operations should have a unique description string and unique media types to select from and other unique parameters (such as query parameters), and thus have separate presentation in swagger-ui. By simply allowing only just schema overloading, there is no way for me to constrain query parameter A to be used with just media type X, and query parameter B to be used with media type Y. These are implemented as two independent operations and it would be nice to be able to model them independently in Swagger.

So I think a proposal i saw elsewhere to annotate the operations themselves (to provide unique keys in the JSON object) would actually be more flexible than other workarounds to the fact that Swagger only allows one POST per path when a REST API may in fact support many independent POST calls.

i.e. something like

``` yaml
paths:
 /a/b/c:
   get: ....
   put: ...
   post.optimize: spec for the optimize operation
   post.merge: spec for the merge operation
```

(Initially discussed in #146)

**Comment by maxlinc** [comment]
Real world example - OpenStack [Server Actions](http://developer.openstack.org/api-ref-compute-v2.html#compute_server-actions). I've been [trying to convert OpenStack WADL to Swagger](https://github.com/rackerlabs/wadl2swagger) and these operations have been a sticking point. There's no real way to describe them right now, because the Swagger doesn't allow path (which is what the WADL uses), nor does it allow "oneOf" or "anyOf" schemas. So I don't see any possible way to describe these services right now.

On the other hand, I want Swagger to define an API, not just describe it. That's the whole reason the working group decided to avoid some of these features in the first place - to avoid ambiguity. If this is going to be done it needs to be done in a way that works for:
- Code generators (like swagger-codegen)
- Stub servers
- Testing tools

It's basically a routing problem: given an HTTP request you should know what operation is being called, and this makes it ambiguous.

**Comment by szhem** [comment]
+100! 

Recently we had to change our api because were unable to define multiple post methods per path.

**Comment by earth2marsh** [comment]
This is a feature of Swagger's opinion on APIs. For better or worse, it presumes that the signature of an operation is the resource path pattern plus the HTTP verb. 

To change this would require an alternate mechanism for defining the signature. If we do pick this up in Swagger.next, then let's consider an overloading of a method with some sort of signature property. I'm not sure that this is something that it SHOULD do, but that is how I would tackle it. (and I agree with what @maxlinc said above)

**Comment by chrisregnier** [comment]
I recently came across the same problem and would love to have the ability to define the API and associate the same path with multiple endpoints that are still unique.

So I'd like to suggest that endpoints are not defined by their paths, but by a user defined key name. More than likely it'll be a domain specific operation name (eg: 'addBookByJson','addBookByBinary' may both have the same path but different endpoints/consumes). Then each endpoint would have a path associated with it instead of 'being' the path directly. Ambiguity is still easy to check based on an equality function of all defined endpoints (which could then be based on implementing platform which might be more applicable rather than defining the equality relation in the swagger API).

There are so many ways I could define an endpoint based on an HTTP request. And if swagger is to be a general tool for documentation, then I would much rather see swagger show its opinions in the form of defaults (or the easy path), NOT by limiting functionality when I need to go outside of the norm to support others or implement new ideas!

**Comment by zeymon** [comment]
+1

**Comment by eli-jordan** [comment]
+1
This would make the tool much more fliexible. I totally agree with @earth2marsh on this point. Allow the user decide what properties of an http request determine its signature.

**Comment by haidaraM** [comment]
+1

**Comment by fabriciorissetto** [comment]
+1

**Comment by eirnym** [comment]
+1

**Comment by who** [comment]
+1

This would be really fantastic:

```
{
    "in": "body",
    "name": "abstractThing",
    "schema": {
        "oneOf": [
            {
                "$ref": "#/definitions/ConcreteThing"
            },
            {
                "$ref": "#/definitions/ConcreteOtherThing"
            }
        ]
    }
}
```

**Comment by JosiasStraesser** [comment]
+1

**Comment by achingbrain** [comment]
+1

**Comment by MendeleyStack** [comment]
+1

**Comment by tom-squires** [comment]
+1

**Comment by jenarros** [comment]
+1

**Comment by DavidBiesack** [comment]
What we are doing in the interim is to use `#tag` in the paths to differentiate these.

  /a/b/c:
    post:
  /a/b/c#merge:
   post:

We have customized Swagger UI to hide these tags and remove them from the Try It/curl
actions

Each of these `post` operations can have their own `parameters/produces/consumes`.

I realize that almost none of the existing Swagger ecosystem will work with this, but we need to move forward, and this is the path we're taking. It was the easiest and lowest impact change to Swagger UI and the swagger documents, and it works fairly well. Our expectation is to be able to transform our use to wherever Swagger.Next goes.

**Comment by gavincornwell** [comment]
+1

**Comment by webron** [comment]
Parent: #574

**Comment by jtaylor10** [comment]
Another possible format for modelling this, not perfect though either. It comes out of the REST principle that you use the verb. Now REST over HTTP forces verb to be GET,POST,etc and they suggest you use an _method parameter or some override parameter to POST operations to indicate your real verb. So with that in mind we could define a paramter with a special type or mark it someway as your verb-override parameter for the specific end point and then change the verb to what you want.

Would look something like this:
`
paths:
   /some/path
      parameters:
           _method:
               type: string
               verb-override: true
      post:
           ....
      copy:
           ....
      merge:
           ....
`

With this, anything that falls outside of the standard HTTP verbs are modelled as POST /some/path?_method=copy.  It could also then be used for those servers that only support PATCH or other standard HTTP but not as well supported verbs through a method override.

**Comment by juan-quiver** [comment]
+1

**Comment by gdyrrahitis** [comment]
+1

**Comment by glen-84** [comment]
@gdyrrahitis Please use [reactions](https://github.com/blog/2119-add-reactions-to-pull-requests-issues-and-comments) now, instead of +1 comments.

**Comment by jbuhacoff** [comment]
I am really hoping that this proposal will be accepted, to index APIs by their logical name instead of path.

I have a set of existing APIs that offer various content-type/accept combinations to accommodate various clients and use cases, and it's not possible to describe them in one openapi 2.0 document.

My temporary solution is to just have a separate openapi document for each use case, so that the inputs and outputs are appropriately bound together.  I had to factor out all of the common definitions to external files in order to minimize the boilerplate.

**Comment by fehguy** [comment]
This particular proposal is not for 3.0, however with https://github.com/OAI/OpenAPI-Specification/pull/798 you can now specify different input and output schemas based on content type.  Please read through that and hopefully that will satisfy your use case.

**Comment by cfineman** [comment]
From my initial reading of this, #146, and #791 this appears to cover what many folks were asking for (might even be usable for some of the protobuf-encoding extensions suggested in #801 ??) 

I'm not up to speed on the "opinionated" aspects of the OpenAPI model but the idea of generic overloading based on parameters makes me a bit uneasy. Having more strongly defined variants for consuming clients is certainly a good thing to strive for but I'm a bit leary of allowing idiosyncrasies of various controller binding models leaking through into a standard (essentially the UI vs implementation argument put forward by @DavidBiesack in #146 )

**Comment by caitsithx** [comment]
+1
what is the final status of this issue?

**Comment by fehguy** [comment]
@caitsithx see my comment https://github.com/OAI/OpenAPI-Specification/issues/182#issuecomment-250600331 for what is happening with 3.0

**Comment by biancama** [comment]
+1

**Comment by ancane** [comment]
+1

**Comment by dalbrekt** [comment]
+1

**Comment by glen-84** [comment]
@ancane @dalbrekt Please use [reactions](https://github.com/blog/2119-add-reactions-to-pull-requests-issues-and-comments) now, instead of +1 comments.

**Comment by jrrevy** [comment]
Adding my 2 cents ;) : 

I have such a case in my project, where I want to use kind of polymorphism around POST method.
I would like to be able to add an item (e.g. a pet) or a collection of items (e.g. multiple pets). By convention _POST /pets/_ means adding **one** item on my pet collection, although my main need is to add a collection of pets.

I should be something like that (maybe wrong)

``` yaml
{
/pets/:
    post:
    "in": "body",
    "name": "PetCollection",
    "schema": {
        "oneOf": [
            {
                "$ref": "#/definitions/Pet"
            },
            {
                type: array
                items:
                  $ref: '#/definitions/Pet'
            }
        ]
    }
```

It could be a fine test to verify that one path leads to multiple operation, depending on content.

I really love the @DavidBiesack workaround. Mine is a bit more ugly but compliant : 
``` yaml
{
/pets/bulk
    post:
    ....
}

The other one is, obviously, to deal with collection every time, even though there is only one item in my collection.
Maybe I'm was wrong aiming to be compliant with convention, don't hesitate to tell me.

It would be related to [#333](https://github.com/OAI/OpenAPI-Specification/issues/333).

**Comment by DavidBiesack** [comment]
@jrrevy glod you like my suggestion... but for OpenAPI 3.0 you can do as you suggest, and use oneOf in your schema. 

What you can't do is make a clear link between passing a single object with schema A (i.e. your singleton) and receiving a response with schema X vs. passing request schema B (i.e. your array input) and getting back schema Y. With oneOf, you can pass { oneOf: [A, B] } and specify the operation returns { oneOf: [X, Y] }. There is no way to express the constraint A -> X, B -> Y (or that A -> Y and B -> X are not possible.) This also applies to query parameters and others aspects that my proposal would solve - if you wish to allow a query parameter P with the collection POST but not allow P with the singleton post, the oneOf schema support does not help.

**Comment by xenohunter** [comment]
@DavidBiesack hello! Could you please share the code where you implemented that trick with hashtags if it's open source? We want to follow your steps as we need our API to process different `requestBody` objects in one path. Thank you in advance!

**Comment by DavidBiesack** [comment]
We don't have any open source/published solution. We use an internal swagger-codegen tool to generate our API doc and we handle the hashes there, but It's not appropriate for open source. Other tools (and there are many) do not support this notation, of course.

**Comment by xenohunter** [comment]
@DavidBiesack OK, thanks anyway! I guess we will implement something on our own then.

**Comment by jeevarajan** [comment]
Any info on how to customize swagger ui to remove #tag from the path in "try it out"/curl actions?

It will be really helping.

**Comment by MikeRalphson** [comment]
@jeevarajan the cleanest way would be by providing a `requestInterceptor` - see the documentation [here](https://github.com/swagger-api/swagger-ui#parameters) and ensure `showMutatedRequest` is set to `true`. Again, please follow up on an issue on the swagger-ui repository if you have any further problems, not here.

**Comment by W7PEA** [comment]
I'm working with two major Integration Platform as a Services (IPaaS) providers who are using Swagger to define the connector to a service to use in a solution.

In using this approach to connect to my company's API, it becomes near impossible to define nice discrete actions on an API that has 'evolved' over time and may offer many similar actions in one endpoint. In an IPaaS Connector's actions, you want nice crisp succent outcomes with the parameters sent limited to just relevant ones. Not POST + a combination of parameters equals one action vs. another action. Yes, our API probably needs refactoring, but in every enterprise these same issues will exist. Given that, the ability to define different POST definitions as outlined above and in the linked issues would be of great benefit.

This issue and the linked issues propose a method.name or method-name approach. I also see other issues to add additional Operation Objects ( #325 TRACE, #1306  SEARCH). This seems to point to the need for a more extensible approach. 

What if instead of specific Operation Objects (GET, PUT, POST...), there was a generic Operation object which had the VERB (maybe METHOD) as part of its schema and then an endpoint could have an array of items Operation. 

```
paths:
    /some/path:
    operations:
        type: array
        items:
            $ref/definitions/operation
            
definitions:
    operation:
        type: object
        properties:
            method:
                type: string
                example: DELETE, GET, POST, PUT, SEARCH, TRACE, ... 
            ....
            All the existing stuff in Operation Object
https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.0.md#operationObject
```

This would allow for extensibility (new Methods) and would also enable the overloading ability others (and I) would like to see.

**Comment by phusuke** [comment]
@DavidBiesack You mentioned that there is an interim solution with tags, 

1.  Is that something that will be restricted to POST or all verbs?
2.  Is that something that is already built out? Because I don't see that support currently.

Thanks

> What we are doing in the interim is to use #tag in the paths to differentiate these.
> 
> /a/b/c:
> post:
> /a/b/c#merge:
> post:
> 
> We have customized Swagger UI to hide these tags and remove them from the Try It/curl
> actions
> 
> Each of these post operations can have their own parameters/produces/consumes.
> 
> I realize that almost none of the existing Swagger ecosystem will work with this, but we need to move forward, and this is the path we're taking. It was the easiest and lowest impact change to Swagger UI and the swagger documents, and it works fairly well. Our expectation is to be able to transform our use to wherever Swagger.Next goes.

**Comment by ma499** [comment]
@xenohunter did you implement anything open source to strip the anchor tags?

**Comment by DavidBiesack** [comment]
I did this work at my previous employer. We did not contribute it back to OpenAPI because it is obvious it would not be accepted. We did not open source this.

**Comment by xenohunter** [comment]
@ma499 no, that turned out to be of lower priority so we have not started yet even.

**Comment by goodwillhacking** [comment]
+1

**Comment by DwayneJengSage** [comment]
+1

We accidentally created two different call patterns for the same verb and URL, and now we're stuck supporting both of them. It would be really nice if there were a way to model this in Swagger.

**Comment by matt212** [comment]
any updates on same how to implement it across !

**Comment by roytanmoy** [comment]
+1

**Comment by cbcye** [comment]
+1

**Comment by membersound** [comment]
I'd love to see `GET` endpoints supporting overload parameters. Eg the parameters-schema could support `oneOf`. Suggestion as follows:
```
{
 "path": {
   "/overloaded:" {
     "get": {
       "parameters": [
	 {
	    "name": "req",
	    "in": "query",
	    "schema": {
		"oneOf": [
		    {
		        "$ref": "#/components/schemas/FirstThing"
		    },
		    {
		        "$ref": "#/components/schemas/SecondThing"
		    }
		]
	    }
	 }
       ]
     }
   }
}
```

**Comment by hkosova** [comment]
@membersound your example is already supported in OpenAPI 3.0.

**Comment by DavidBiesack** [comment]
It goes beyond this, @whitlockjc - i.e. not just input/output schemas, but extends to separate sets of query/header parameters (in addition to body params). At present, even tho OpenAPI 3.x lets you "partition" schemas based on media types, one must bundle all header/query parameters together with the one operation. We really want to be able let other signature items (not just path pattern + verb) to be used to "identify" or route a request and have separate operations for each.

**Comment by bkeeler-redox** [comment]
Wow, this is a big gap in OpenAPI, so it's quite surprising to see how long this discussion has been ongoing.

**Comment by fabiocastilhos** [comment]
+1

**Comment by Avkashhirpara** [comment]
I am facing the same issue. I want to document same path with multiple parameters. I have checked anyOf,OneOf, options but the it doesn't server the purpose. Need any other identifier other then Http methods + path. 

if there is any update or alternative on the same, It would be helpful.

**Comment by travishaagen** [comment]
The long-standing work-around is to add a unique URL-fragment to the end of the path. However, if your OpenAPI JSON description is being auto-generated by some library that doesn't support this use-case, then you either need to find a different library or write your own code to generate the OpenAPI JSON. You could also hand-write the OpenAPI JSON :)

**Comment by dahei** [comment]
> What we are doing in the interim is to use `#tag` in the paths to differentiate these.
> 
> /a/b/c:
> post:
> /a/b/c#merge:
> post:

To visually "hide" this `#tag` path param you can also use an empty unicode character like "Hairspace" (U+200A) - see https://emptycharacter.com/ for examples.

**Comment by saiya** [comment]
I encountered same problem when I am writing OpenAPI spec of my OAuth2 server. 

My workaround is to use path parameter like below (rather than [`#tag`](https://github.com/OAI/OpenAPI-Specification/issues/182#issuecomment-164037595) way):

```javascript
{
  "paths": {
    "/oauth/{code_grant_endpoint}": {
      "post": {
        "operationId": "codeGrant",
        "parameters": [
          { "name": "code_grant_endpoint", "in": "path", "required": true, "schema": { "type": "string", "enum": [ "token" ] } },
          // .... and other parameters ...
        ],
        // ...
      },
     "/oauth/{client_cred_endpoint}": {
      "post": {
        "operationId": "clientCredentialsGrant",
        "parameters": [
          { "name": "client_cred_endpoint", "in": "path", "required": true, "schema": { "type": "string", "enum": [ "token" ] } },
          // .... and other parameters ...
        ],
        // ...
      },
      // ...
    },
}
```

In above example, I defined two operations "codeGrant" and "clientCredentialsGrant".

Path of both operations are exactly same: `/oauth/token`. 

Because of the path parameters in definition (`{code_grant_endpoint}`, `{client_cred_endpoint}`), it does not conflict in OpenAPI definition. With using `enum` schema restriction, the path is not variable actually.

This way works without any modfication or postprocess in OpenAPI definition consumer side (e.g. openapi-generator).

**Comment by hkosova** [comment]
@saiya 
> Because of the path parameters in definition ({code_grant_endpoint}, {client_cred_endpoint}), it does not conflict in OpenAPI definition.

Actually the `/oauth/{code_grant_endpoint}` and `/oauth/{client_cred_endpoint}` paths are [considered identical](https://spec.openapis.org/oas/v3.1.0.html#patterned-fields) and therefore invalid.

From the spec (emphasis mine):

> Templated paths **with the same hierarchy but different templated names** MUST NOT exist as they are identical.

**Comment by saiya** [comment]
> paths are considered identical and therefore invalid.

Oh... thank you for comment.

It seems work in my case, but now I understand it is implementation specific behavior...

Sad for the restriction.

**Comment by btodts** [comment]
+1

**Comment by vellala2000** [comment]
+1

**Comment by jon-jon-jon** [comment]
My 5 cent and solution:
Instead of defining your query parameters (normal way), just hard-code your query id in the URL and set its value as "path" type. It is not perfect but it's working fine =) You still use component, description and other Swagger's functionality to be more friendly ;)

```yaml
 /myUrl/myRoute/?queryPara1={query1}?queryPara2={query2}:
    get:
      parameters:
        - in: path
          name: query1
          schema:
            type: string
        - in: path
          name: query2
          schema:
            type: string

  /myUrl/myRoute/?queryPara3={query3}?queryPara4={query4}:
    get:
      parameters:
        - in: path
          name: query3
          schema:
            type: string
        - in: path
          name: query4
          schema:
            type: string
```

**Comment by DavidBiesack** [comment]
@jon-jon-jon interesting... OpenAPI does not seem to restrict the value of a path to be just `relative-part` of [RFC 7230](https://datatracker.ietf.org/doc/html/rfc7230#section-2.7), at least not formally ("A relative path to an individual endpoint") , i.e. the OpenAPI path does not specifically exclude the `[ "?" query ] [ "#" fragment ]` portion.  That's up to the OAS maintainers to address. So be wary; they may impose such a restriction in the future.
Also keep in mind that tools that consume this OpenAPI (SDK code gen, etc.) would likely get confused by this, as these parameters are not really path parameters but query parameters, and libraries to generate paths or parse paths may not work. Also, query parameter order does not matter (except for array params), so your API won't support valid calls with different order `/myUrl/myRoute/?queryPara4={query4}&queryPara3={query3}` which would be expected with properly defined query parameters.
Finally, with OAS, parameters with `in: path` must also be `required: true`, so if "it's working fine", that is only in the sense that perhaps your code supports it, but the actual OAS document would be invalid.

**Comment by karenetheridge** [comment]
"it's working fine" suggests that the specification needs to be more clear here (https://spec.openapis.org/oas/v3.1.0#patterned-fields). In order to match the request's path against all the path-item properties, we need to clearly define just what the "path" is. It can't be true that the path both contains and DOESN'T contain the query parameters.

Regarding fragments, I think we can be a little more certain, because fragments are not supposed to be handled by servers at all - it's purely a client-side thing. I would certainly regard any attempt by openapi tooling to specify or check the fragment to be an error.

**Comment by peteraritchie** [comment]
> Regarding fragments, I think we can be a little more certain, because fragments are not supposed to be handled by servers at all - it's purely a client-side thing. I would certainly regard any attempt by openapi tooling to specify or check the fragment to be an error.

User agents aren't even expected to send the fragment in the request to the server.

**Comment by matt-j-stevenson** [comment]
> What we are doing in the interim is to use `#tag` in the paths to differentiate these.
> 
> /a/b/c: post: /a/b/c#merge: post:
> 
> We have customized Swagger UI to hide these tags and remove them from the Try It/curl actions
> 
> Each of these `post` operations can have their own `parameters/produces/consumes`.
> 
> I realize that almost none of the existing Swagger ecosystem will work with this, but we need to move forward, and this is the path we're taking. It was the easiest and lowest impact change to Swagger UI and the swagger documents, and it works fairly well. Our expectation is to be able to transform our use to wherever Swagger.Next goes.

This was some time ago, but this thread has been a good source of truth, is there chance you could tell me how you removed the tag from the 'Try it out' interface?

**Comment by handrews** [comment]
The [Moonwalk (OAS 4)](https://github.com/OAI/moonwalk/discussions) proposal has [affirmed a principle of operation signatures](https://www.openapis.org/blog/2023/12/06/openapi-moonwalk-2024) which will address this use case 🎉 

As this cannot be done within the 3.x line, please follow the Moonwalk repo discussions for further developments.

**Comment by handrews** [comment]
@eirnym Swagger is not the only product in the OpenAPI ecosystem.  Other major tooling vendors started adding 3.1 support as early as 2 months after 3.1.0's publication.  Swagger's late support just ensured that new OAS users looking for modern technologies chose other vendors.  From an ecosystem point of view, it's actually better to have multiple competing major vendors rather than a single dominant one.

Adoption of 3.1 continues to grow, including in larger organizations less likely to be on the bleeding edge.

OAS 3.0.0 was released in 2017.  When we published OAS 3.1.0 four years later, OAS 3 still had not really supplanted OAS 2.  But by now it has.  These things always take more time than anyone would like.  

That said, we are already looking at ways to make Moonwalk more adoptable.

**Comment by eirnym** [comment]
I totally agree that swagger is only one of tools to be used for OAS. However, I've observed in few enterprise companies that "We can't use OAS 3.1 because of Swagger and we don't accept any other third party tool". 

What do you propose for developers to say teams responsible for API in a company to increase popularity of Moonwalk (as a next OAS generation)?

**Comment by handrews** [comment]
@eirnym This is a complex topic that probably belongs somewhere other than comments on a random closed/moved issue.  But I'll take a shot at it here- this is something that interests me very much, along with ensuring that we have a series of _small_ but useful 3.x releases to both deliver features for users and tooling vendors who are just now getting to 3.1 and make the  migration to Moonwalk more smooth when it becomes possible.

My personal view (and while as a maintainer I have some write privileges in this repo, I am not in any way speaking as a representative of the project's official position), is that the OpenAPI Initiative has lost the trust of at least significant part of its community.  This due to several things, all of which are understandable individually but have added up to a difficult state of affairs:

* 3.1 was really a _major_ release, and in fact quite a few people wanted to call it 4.0, but the marketing faction won out and we ditched semver and called it 3.1 on the grounds that, technically, the breakage was just that we replaced `type: X, nullable: true` with `type: [X, "null"]`.  Which was true.  But to really add full JSON Schema support was a huge effort, and I think no one person involved quite understood the entire impact at the time.
* We've never actually published  a minor release, so there's no trust (or, at this point, even expectation) that we'll use minor releases to meet the short-to-medium term needs of the community by releasing compatible improvements that tooling vendors can add to their existing code-bases without much effort.
* We've stopped even publishing patch releases - 3.0.3 was published before 3.1.0.  There's a ridiculous number of issues filed that [are basically just asking for clarifications](https://github.com/OAI/OpenAPI-Specification/issues?q=is%3Aissue+is%3Aopen+label%3Aclarification) that could have been regularly released as patch releases that would only make the lives of tooling vendors easier by removing guesswork in certain areas of the spec.
* We've created [other places](https://learn.openapis.org) where we can address some things, like the [huge demand for way more examples than would fit in the spec](https://github.com/OAI/OpenAPI-Specification/issues?q=is%3Aissue+is%3Aopen+label%Aexamples), but have not been able to promote or expand it as much as I think we need to.
* The issue backlog became absurd.  I think we had over 600 issues open before Lorna and I started closing things out and/or moving them over to Moonwalk.  Questions were left unanswered for years, there was no indication of which things might be worked on or when anything might reach actual users.  There are also too many channels and not enough attention - I just recently found out that discussions are also enabled on this repo.

I think a lot of this had to do with the pandemic.  Several people, including myself, became unable to focus on these projects.  There have also been some weird things that have been logistically challenging due to unavailability of people who worked on those things in the past, and lack of understanding of how and why things are the way they are.  The broken links on the rendered Markdown in GitHub would be the most glaring example.  There were also people who felt that with 3.1.0 out, there shouldn't be more 3.0.x releases.  And there was no point in doing 3.1.1 when 3.1.0 wasn't implemented yet.

Last year things were still moving slowly and sporadically as I got involved again (I was moving pretty slowly and sporadically, too).  But there is a huge burst of energy from new and returning people as we kick off 2024, and we are [actively working to put our house in order](https://github.com/OAI/OpenAPI-Specification/issues/3516#).

This is starting with the expansion of the Technical Steering Committee so that we have a full roster of active TSC members.  Once we have our governance revived, we'll be able to make more headway on that list of process issues.

We've changed the agenda for the weekly Thursday calls to block out time for process and governance at the start of the meeting to make sure we make progress.  We've also split Moonwalk out into a Tuesday meeting, leaving the Thursday meeting to focus on 3.x plus governance (we've also considered splitting out the governance work, but at the moment it feels like that's not needed and would be too many meetings).  We're also talking about finding some meeting times that suit more time zones and maybe rotating the time slot a bit.

Also in that New Year's Cleaning meta-issue, you'll see issues about defining our minor and patch release strategies for 3.x(.y).  You'll see that I've proposed an overall philosophy of "smoothing the path to Moonwalk."  I'm not the only person advocating for 3.x – I think @lornajane first brought up the need to keep our focus on it and not just rush off to Moonwalk, and she has continued to advocate for it in the calls.

I think we have to rebuild trust by clarifying as much as we can for vendors and users in 3.0.x and 3.1.x patch releases (since the majority of users are still on 3.0), and by releasing small, incremental, _valuable_ 3.x releases so that end-users see benefits and signs that their needs are being listened to well before Moonwalk support becomes widespread.  Where possible, we should backport Moonwalk features to 3.x to increase convergence (like Python 2.x and 3.x releases when they overlapped).  I'm currently going a bit overboard adding new labels to the remaining backlog to figure out the patterns in requests that could help us define coherent and valuable 3.x(.y) releases.

I have some other ideas about making the path to moonwalk less of a hug gap to jump across, but I haven't floated them with anyone yet and they're a little too half-baked to write about publicly at this point.  But I plan to be in both the Tuesday 4.x and Thursday 3.x calls on a regular basis, pushing for something that is not just "release a de-facto major version that requires a complete architectural re-write every 3 to 4 years."

Moonwalk is exciting, but my _personal_ opinion is that we need to tend to the 3.x users, and show that it is worthwhile to move to 3.1 because they will be rewarded with a steady stream of incremental 3.x releases that will make the transition to Moonwalk as smooth as possible when the time comes.
